### PR TITLE
Implement `__slots__` with `member_descriptor`

### PR DESCRIPTION
**Pull request by moreal**
This pull request tries to implement `__slots__` feature with `member_descriptor`.

This pull request doesn't contain `mebmer_descriptor.slots.descr_get` yet and it'll be implemented in the next pull request because I worried that this pull request becomes too big to review.

### CODE REVIEW

File: vm/src/builtins/descriptor.rs [hunk 1]
```diff
@@ -0,0 +1,134 @@
+use rustpython_common::lock::PyRwLock;
+
```
Comment (youknowone) [inline vm/src/builtins/descriptor.rs:2]:
```suggestion
```

File: vm/src/builtins/descriptor.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+use crate::function::Either;
+use crate::types::{Constructor, GetDescriptor, Unconstructible};
+use crate::{AsObject, Context, Py, PyObjectRef, PyPayload, PyRef, PyResult, VirtualMachine};
+
+#[derive(Debug)]
+pub struct DescrObject {
+    pub typ: PyTypeRef,
+    pub name: String,
+    pub qualname: PyRwLock<Option<String>>,
+}
```
Comment (youknowone) [inline vm/src/builtins/descriptor.rs:14]: 👍

File: vm/src/builtins/descriptor.rs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+#[derive(Debug)]
+pub enum MemberKind {
+    ObjectEx = 16,
+}
+
+pub struct MemberDef {
+    pub name: String,
+    pub kind: MemberKind,
+    pub getter_or_offset: Either<fn(PyObjectRef, &VirtualMachine) -> PyResult, usize>,
```
Comment (youknowone) [inline vm/src/builtins/descriptor.rs:24]: Using a dedicated enum type will be better to adding utility functions.

```rust
enum MemberGetter {
    Getter(fn(&VirtualMachine, PyObjectRef) -> PyResult),
    Offset(usize),
}
```
Swapping parameter orders in Getter will be helpful to have same convention with `PyGetterFnuc`.

File: vm/src/builtins/descriptor.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+impl PyPayload for MemberDescrObject {
+    fn class(vm: &VirtualMachine) -> &'static Py<PyType> {
+        vm.ctx.types.member_descriptor_type
+    }
+}
+
+fn calculate_qualname(descr: &DescrObject, vm: &VirtualMachine) -> PyResult<Option<String>> {
+    let type_qualname = vm.get_attribute_opt(descr.typ.to_owned().into(), "__qualname__")?;
```
Comment (youknowone) [inline vm/src/builtins/descriptor.rs:52]:
```suggestion
    let type_qualname = if let Some(qualname) = vm.get_attribute_opt(descr.typ.to_owned().into(), "__qualname__")? {
        qualname
    } else {
        return Ok(None);
    };
```

File: vm/src/builtins/descriptor.rs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    match type_qualname {
+        None => Ok(None),
+        Some(obj) => match obj.downcast::<PyStr>() {
+            Ok(str) => Ok(Some(format!("{}.{}", str, descr.name))),
+            Err(_) => Err(vm.new_type_error(
+                "<descriptor>.__objclass__.__qualname__ is not a unicode object".to_owned(),
+            )),
+        },
+    }
```
Comment (youknowone) [inline vm/src/builtins/descriptor.rs:61]:
```suggestion
    let str = obj.downcast::<PyStr>().map_err(|_| vm.new_type_error(
                "<descriptor>.__objclass__.__qualname__ is not a unicode object".to_owned(),
    ))?;
    Ok(Some(format!("{}.{}", str, descr.name))))
```

File: vm/src/builtins/descriptor.rs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        zelf.member.doc.to_owned()
+    }
+
+    #[pyproperty(magic)]
+    fn qualname(&self, vm: &VirtualMachine) -> PyResult<Option<String>> {
+        if self.common.qualname.read().is_none() {
+            *self.common.qualname.write() = calculate_qualname(&self.common, vm)?;
+        }
+
+        Ok(self.common.qualname.read().to_owned())
```
Comment (youknowone) [inline vm/src/builtins/descriptor.rs:86]: reduce lock once when it is some
```suggestion
        let qualname = self.common.qualname.read();
        Ok(if qualname.is_none() {
            drop(qualname);
            let calculated = calculate_qualname(&self.common, vm)?;
            *self.common.qualname.write() = calculated;
            calculated
        } else {
            qualname.to_owned()
        })
```

File: vm/src/builtins/descriptor.rs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        _cls: Option<PyObjectRef>,
+        vm: &VirtualMachine,
+    ) -> PyResult {
+        match obj {
+            Some(x) => {
+                let zelf = Self::_zelf(zelf, vm)?;
+                match zelf.member.getter_or_offset {
+                    Either::A(getter) => (getter)(x, vm),
+                    Either::B(offset) => get_slot_from_object(x, offset, &zelf.member, vm),
+                }
```
Comment (youknowone) [inline vm/src/builtins/descriptor.rs:124]: When `zelf.member.getter` is `MemberGetter`,

```suggestion
                zelf.member.getter.get(x, vm)
```

**Comment by moreal** [comment]
Currently, this pull request add 381 lines with 12 commits and it needs to implement setter also. And I saw passing CI with success before applying reviews so I expect it will pass now too. So I thought it can stop working on this pull request in current step and should implement the setter in the next PR. 🤔 

I'll make this pull request as `Ready for review`

**Review by youknowone — APPROVED** [review]
`PyType::slot_new` is now huge chunk of side effects. I want to break them down but it doesn't look easy.

**Pull request merged**